syntax = "proto3";

package raccoon.agent.v1;

option java_multiple_files = true;
option java_package = "com.raccoon.agent.v1";

import "google/protobuf/empty.proto";
import "google/protobuf/timestamp.proto";
import "google/protobuf/struct.proto";

// AgentService handles LLM orchestration and agent execution.
service AgentService {
  // ExecuteAgent runs an agent turn with streaming response events.
  // Default deadline: 60s per turn (configurable by deployment).
  rpc ExecuteAgent (AgentRequest) returns (stream AgentResponse);

  // GetAgentConfig retrieves the current configuration for an agent.
  rpc GetAgentConfig (AgentConfigRequest) returns (AgentConfig);

  // ValidateTools checks whether tool configurations are valid.
  rpc ValidateTools (ValidateToolsRequest) returns (ValidateToolsResponse);
}

// SandboxService manages E2B sandbox lifecycle and code execution.
service SandboxService {
  // CreateSandbox provisions a new E2B sandbox for a conversation.
  rpc CreateSandbox (CreateSandboxRequest) returns (SandboxInfo);

  // ExecuteCode runs code inside an existing sandbox with streaming output.
  // Default deadline: 45s per execution chunk.
  rpc ExecuteCode (ExecuteCodeRequest) returns (stream ExecutionOutput);

  // UploadFile sends a file into a sandbox.
  rpc UploadFile (UploadFileRequest) returns (UploadFileResponse);

  // DestroySandbox tears down a sandbox and releases resources.
  rpc DestroySandbox (DestroySandboxRequest) returns (google.protobuf.Empty);
}

// ---------------------------------------------------------------------------
// AgentService messages
// ---------------------------------------------------------------------------

message Message {
  string role = 1;       // "user", "assistant", "system", "tool"
  string content = 2;
  string message_id = 3;
  google.protobuf.Timestamp timestamp = 4;
  repeated Attachment attachments = 5;
}

message Attachment {
  string filename = 1;
  string mime_type = 2;
  bytes data = 3;
  string url = 4;        // Alternative: reference by URL instead of inline bytes
}

message AgentRequest {
  string conversation_id = 1;
  string agent_id = 2;
  repeated Message messages = 3;
  AgentConfig config = 4;
  string user_api_key = 5;   // BYOK â€” empty means use platform credits
  string request_id = 6;     // Idempotency key
}

// AgentResponse is a stream of events emitted during agent execution.
message AgentResponse {
  oneof event {
    TokenEvent token = 1;
    StatusEvent status = 2;
    ToolCallEvent tool_call = 3;
    ToolResultEvent tool_result = 4;
    CodeBlockEvent code_block = 5;
    ErrorEvent error = 6;
    ApprovalRequestEvent approval_request = 7;
    CompleteEvent complete = 8;
  }
}

// TokenEvent represents a single streamed text token.
message TokenEvent {
  string text = 1;
}

// StatusEvent carries a fun developer-humor status message.
message StatusEvent {
  string message = 1;     // e.g. "consulting the raccoon council..."
  string category = 2;    // e.g. "thinking", "coding", "deploying"
}

// ToolCallEvent is emitted when the agent invokes a tool.
message ToolCallEvent {
  string tool_call_id = 1;
  string tool_name = 2;
  google.protobuf.Struct arguments = 3;
}

// ToolResultEvent carries the result of a tool invocation.
message ToolResultEvent {
  string tool_call_id = 1;
  string tool_name = 2;
  bool success = 3;
  string output = 4;
  string error_message = 5;
}

// CodeBlockEvent is emitted when the agent generates a code artifact.
message CodeBlockEvent {
  string language = 1;
  string code = 2;
  string filename = 3;     // Optional filename hint
}

// ErrorEvent represents a terminal or recoverable error.
message ErrorEvent {
  string code = 1;          // Stable error code (e.g. "deadline_exceeded")
  string message = 2;
  bool recoverable = 3;
}

// ApprovalRequestEvent is emitted before executing a sensitive tool action.
// The client MUST respond with an approval decision before execution continues.
message ApprovalRequestEvent {
  string approval_id = 1;
  string tool_name = 2;
  google.protobuf.Struct arguments = 3;
  string reason = 4;        // Human-readable explanation of why approval is needed
}

// CompleteEvent signals the agent turn has finished.
message CompleteEvent {
  int32 input_tokens = 1;
  int32 output_tokens = 2;
  string model = 3;
  string stop_reason = 4;   // "end_turn", "max_tokens", "tool_use"
}

// ---------------------------------------------------------------------------
// Agent configuration
// ---------------------------------------------------------------------------

message AgentConfigRequest {
  string agent_id = 1;
}

message AgentConfig {
  string agent_id = 1;
  string system_prompt = 2;
  string model = 3;
  double temperature = 4;
  int32 max_tokens = 5;
  repeated ToolConfig tools = 6;
  string visibility = 7;     // "public", "private", "unlisted"
}

message ToolConfig {
  string name = 1;
  string description = 2;
  google.protobuf.Struct input_schema = 3;
  bool requires_approval = 4;
}

// ---------------------------------------------------------------------------
// Tool validation
// ---------------------------------------------------------------------------

message ValidateToolsRequest {
  repeated ToolConfig tools = 1;
}

message ValidateToolsResponse {
  bool valid = 1;
  repeated ToolValidationError errors = 2;
}

message ToolValidationError {
  string tool_name = 1;
  string error = 2;
}

// ---------------------------------------------------------------------------
// SandboxService messages
// ---------------------------------------------------------------------------

message CreateSandboxRequest {
  string conversation_id = 1;
  string template = 2;        // e.g. "python", "node", "react"
  SandboxLimits limits = 3;
}

message SandboxLimits {
  int32 max_cpu = 1;            // Number of CPU cores
  int32 max_memory_mb = 2;      // Memory limit in MB
  int32 timeout_seconds = 3;    // Maximum sandbox lifetime
}

message SandboxInfo {
  string sandbox_id = 1;
  string conversation_id = 2;
  string template = 3;
  string status = 4;            // "creating", "ready", "running", "terminated"
  string preview_url = 5;       // Live preview URL for page generation
  google.protobuf.Timestamp created_at = 6;
  SandboxLimits limits = 7;
}

message ExecuteCodeRequest {
  string sandbox_id = 1;
  string code = 2;
  string language = 3;          // "python", "javascript", "bash", etc.
  int32 timeout_seconds = 4;    // Per-execution timeout (default: 45s)
}

// ExecutionOutput is a stream of output events from code execution.
message ExecutionOutput {
  oneof output {
    string stdout = 1;
    string stderr = 2;
    ExecutionResult result = 3;
    ExecutionError error = 4;
  }
}

message ExecutionResult {
  int32 exit_code = 1;
  string output = 2;
  repeated OutputFile files = 3;   // Files produced by execution
}

message ExecutionError {
  string code = 1;
  string message = 2;
}

message OutputFile {
  string path = 1;
  string mime_type = 2;
  bytes data = 3;
}

message UploadFileRequest {
  string sandbox_id = 1;
  string path = 2;              // Destination path in sandbox
  bytes data = 3;
  string mime_type = 4;
}

message UploadFileResponse {
  string path = 1;
  int64 size_bytes = 2;
}

message DestroySandboxRequest {
  string sandbox_id = 1;
}
